### DISCUSSION TITLE
[Bug Report] [bug] Mobile Base yaw not updating xy translation

### DISCUSSION THREAD

**Discussion by exaFLOPs26** [isaac-sim/IsaacLab#2664] [Q&A]
### Describe the bug

Mobile bases with no distinct wheels (e.g., ridgeback Franka) have a hierarchical joint setup: world -> dummy_joint_x -> dummy_joint_y -> mobile_base. So even if I rotate the yaw, xy is not affected. So it goes like this.

https://github.com/user-attachments/assets/555108f9-3973-48f6-9c68-efc633ac2450

### Steps to reproduce

1. First I change the x_joint and it went forward and backward direction.
2. As I rotate the yaw for around 90 degree x,y should have changed.
3. Last I change the y_joint that the direction should have been forward and backwards but still sideways, which means that the yaw rotation is not updating the xy translation. And it is obvice by the hierarchical joint structure.

### What I did first

Change it to yaw-x-y joint hierarchical structure. But now it rotates in a circle. This is obvice even I change the xy translation, again in the opposite way, xy translation is not updating yaw rotation. So the center of yaw rotation is fixed at the first time.

### What I am doing

I use simple linear algebra to calculate the xy translation with yaw rotation. But I can't  move and rotate at the same time. And this is a big bottleneck on Imitation Learning and RL for mobile ones. 

<img width="461" alt="Image" src="https://github.com/user-attachments/assets/a2c2145c-2c48-4b4c-b50d-9b521dc428de" />

### System Info

Describe the characteristic of your environment:

- Isaac Sim Version: 4.5.0
- OS: Ubuntu 22.04
- GPU: Nvidia L40S

<img width="301" alt="Image" src="https://github.com/user-attachments/assets/42f890fa-aec3-4f9e-a788-41e4f64031ab" />

### Additional context

- To do Sim2Real Real2Sim is important. How can we tune the simulation to make it stablized? Tried gain tuner, but at the end of the day, I did it manually.

- I remember a speech by Eric in office hour session, that what we need to do is not make the simulation same as the real world, but make it cover it. Are there any tips on this also?

![Image](https://github.com/user-attachments/assets/ea9526a0-07f7-43eb-8808-3e8d99639c89)

**Comment by RandomOakForest** [comment]
Thank you for posting this. Interesting catch, the team will review this. It would be great if you can start new _discussions_ on each of your questions in "additional context", adding more details or elaborating on a specific task of interest.

**Comment by exaFLOPs26** [comment]
Thanks! One more thing is that if I do ik absolute control for these kind of mobile manipulator(without distinct wheel like ridgeback_franka) it occurs a problem. 

I fixed the 'source/isaaclab/isaaclab/envs/mdp/task_space_actions.py' to make it work! If interested, I can share it.

**Comment by RandomOakForest** [comment]
That would be great. Please add a PR for the team to review.

**Comment by RandomOakForest** [comment]
Following up, it seems the video is showing the expected behavior as you update parameters. In particular, we don't think this is expected: "yaw for around 90 degree x,y should have changed."

**Comment by exaFLOPs26** [comment]
First I control the x joint and next I controled the Y joint. As I rotate 90 degree, wouldn't the x and y head the same direction?

I think my demo is quite overcomplicated, but the problem I am having is that the robot is not going forward toward the heading direction but just goint along the fixed world x axis even the yaw is rotated.

**Comment by exaFLOPs26** [comment]
https://forums.developer.nvidia.com/t/moving-ridgebackfranka-robot-with-articulations-does-not-change-the-pose-of-the-robot/242112/8

**Comment by exaFLOPs26** [comment]
I found out that others are having the same issue. How can we handle this?

**Comment by exaFLOPs26** [comment]
@RandomOakForest Not to rush you, but this is still a bug to fix right?

**Comment by RandomOakForest** [comment]
Following up, I'll move this post to our _Discussions_ section. Here is a summary that may help with possible next steps to try. To achieve proper movement in local x/y directions after rotating the mobile base, you need to **transform control inputs from the local frame to world frame** using the robot's current orientation. Here's the technical breakdown:

### Core Requirements

1. **Frame Transformation**
    - After a 90° yaw rotation, the mobile base's local axes no longer align with world axes
    - Movement commands must be rotated using current orientation data
    - Use `quat_apply()` [^5] or `transform_points()` [^3] for coordinate transformation:

```python
# Local forward direction (x-axis)
local_vec = torch.tensor([1.0, 0.0, 0.0])  

# Get current orientation quaternion (w,x,y,z format)
quat = mobile_base.get_orientation()  

# Transform to world frame
world_vec = quat_apply(quat, local_vec) [^3][^5]
```

2. **Hierarchical Joint Control**
    - Dummy joints operate in parent frame space [^2]
    - After transformation, apply calculated world vectors to joints:

```python
# For velocity control
dummy_joint_x.set_velocity(world_vec[^0] * speed)
dummy_joint_y.set_velocity(world_vec[^1] * speed)
```


### Implementation Checklist

**1. Orientation Tracking**

- Continuously monitor mobile base orientation using `get_orientation()`

**2. Command Transformation**

- Convert desired local movement vectors using current quaternion

**3. Joint Actuation**

- Apply transformed vectors to dummy prismatic joints [^2]
- For position control, use cumulative offsets:

```python
current_x = dummy_joint_x.get_position()
current_y = dummy_joint_y.get_position()
new_x = current_x + world_vec[^0] * step_size
new_y = current_y + world_vec[^1] * step_size
dummy_joint_x.set_position(new_x)
dummy_joint_y.set_position(new_y)
```


### Common Pitfalls

- **Frame Stacking**: Ensure joint hierarchy matches your coordinate transformation order (world→X→Y→base) [^2]
- **Quaternion Convention**: Verify quaternion storage format (w,x,y,z vs x,y,z,w) in your USD setup [^3]

The critical insight is that **dummy joints don't automatically account for child rotations** - you must manually transform control vectors using the mobile base's current orientation before applying them to the X/Y joints[^2][^3][^5].

_References_

[^1]: https://forums.developer.nvidia.com/t/simplifying-joint-position-and-rotation/264606

[^2]: https://forums.developer.nvidia.com/t/moving-ridgebackfranka-robot-with-articulations-does-not-change-the-pose-of-the-robot/242112

[^3]: https://isaac-sim.github.io/IsaacLab/v2.0.2/source/api/lab/isaaclab.utils.html

[^4]: https://forums.developer.nvidia.com/t/how-to-rotate-a-revolute-joint-smoothly-from-pi-to-pi/279572

[^5]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.utils.html

[^6]: https://isaac-sim.github.io/IsaacLab/main/source/refs/changelog.html

[^7]: https://uan.csail.mit.edu/rsc/paper.pdf

[^8]: https://developer.nvidia.com/blog/closing-the-sim-to-real-gap-training-spot-quadruped-locomotion-with-nvidia-isaac-lab/

[^9]: https://github.com/isaac-sim/IsaacLab/discussions/2362

[^10]: https://forums.developer.nvidia.com/t/is-there-a-way-to-find-out-the-local-coordinate-rotation-of-an-object-in-isaac-lab/302462

[^11]: https://isaac-sim.github.io/IsaacLab/main/source/overview/teleop_imitation.html

[^12]: https://docs.robotsfan.com/isaaclab_official/v2.1.0/_modules/isaaclab/utils/math.html

[^13]: https://arxiv.org/html/2502.01546v1

[^14]: https://github.com/isaac-sim/IsaacLab/issues/1436

[^15]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.devices.html

[^16]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.assets.html

**Comment by exaFLOPs26** [comment]
I see! Thanks for the detailed reply. 
Sadly, I just made an omniwheel inside and made it work but when I need to do this I will definitely refer to your suggestions. Thanks again!
